getinsomnia/insomnia
# Allow changing the protofile for a request

The selected method, and the body contents are reset when changing the proto file linked to a request. This PR also adds some edge case handling.

![2020-11-11 17 22 31](https://user-images.githubusercontent.com/4312346/98765266-9c151180-2442-11eb-8bda-f73b8fea320b.gif)

<details>
<summary>Proto file deletion edge case</summary>

@nijikokun there are edge cases that arise on deletion of a proto file, evident especially through this PR. When a request is activated or a message sent, the proto file is loaded again from the contents in the database. But if a proto file has been deleted, those contents are lost, and we can't fetch all of the information required in order to render the UI correctly (the tab title and method drop-downs, for example). In this case, when the protofile for the active request is deleted, the UI is not reset immediately, but on navigating away and returning to the request, the UI will be reset.

![2020-11-11 17 22 49](https://user-images.githubusercontent.com/4312346/98765606-64f33000-2443-11eb-8c59-c332ec187692.gif)

</details>

See collapsed comment above for some context about the following changes. The behavior was discussed with @nijikokun before being implemented.

Previously, the methods for a protofile linked to a request were saved in local state, meaning each time the user switched between requests in the sidebar, the methods would be reloaded from the protofile. Because of this, if the protofile for a given request has been deleted and a user tries to activate it, no UI can be rendered because the data no longer exists. Storing the method information itself against a request allows the UX to not be immediately broken (only on the next restart), and has a side-benefit of allowing the business logic and React logic to be more performant.

With this change, an additional action to invalidate the methods was introduced. Invalidating tells React that the next time you need methods, reload them instead of loading from state. Currently, methods are invalidated when a protofile is changed, which is the primary _feature_ introduced by this PR.

I suggest starting your review from the `grpc-reducer.js` file.

QA:
- smoke testing of requests in general: canceling, multiple streams, can make requests etc
- when switching methods, the UI should clear
- when changing protofile, the UI should clear
- if the protofile for a request is deleted, the UI (previous request/response information) should NOT clear
  - it will clear on restart, because it is only stored in memory

Note, if you were to upload the _same_ proto file twice, the application will treat it as two unique files as there is no de-duplication in place.

Fixes INS-211, Fixes INS-272
